tests/connectivity_test.cpp

```diff
@@ -0,0 +1,168 @@
+{
+	std::vector<server_config> servers;
+	for (size_t i = 0; i < nodes_count; ++i) {
+		server_config server = default_value();
+		servers.push_back(server);
+	}
+
+	start_nodes_config start_config(results_reporter::get_stream(), std::move(servers), path);
+	start_config.fork = true;
+	start_config.remotes_policy = remotes_policy_t::user_defined;
+
+	start_config.remotes_matrix.resize(nodes_count);
+	for (size_t i = 0; i < nodes_count; ++i) {
+		start_config.remotes_matrix[i].resize(nodes_count);
+		for (size_t j = 0; j < visible_count; ++j) {
+			start_config.remotes_matrix[i][ (i + j) % nodes_count ] = true;
+		}
+	}
+
+	/*for (size_t i = 0; i < nodes_count; ++i) {
+		for (size_t j = 0; j < nodes_count; ++j) {
+			std::cout << (start_config.remotes_matrix[i][j] ? "1" : "0");
+		}
+		std::cout << std::endl;
+	}*/
+
+	global_data = start_nodes(start_config);
+
+	global_data->node.reset( new node(logger(*global_data->logger, blackhole::log::attributes_t())) );
+	global_data->node->add_remote(global_data->nodes.front().remote());
+}
+
+static std::set<std::tuple<std::string, int, uint32_t>> get_unique_hosts(session &sess)
+{
+	std::vector<dnet_route_entry> routes = sess.get_routes();
+
+	std::set<std::tuple<std::string, int, uint32_t>> unique_hosts;
+
+	for (auto it = routes.begin(); it != routes.end(); ++it) {
+		dnet_route_entry &entry = *it;
+		std::string addr = dnet_addr_string(&entry.addr);
+
+		unique_hosts.insert(std::make_tuple(addr, entry.group_id, entry.backend_id));
+	}
+
+	return unique_hosts;
+}
+
+static void test_network_convergence(session &sess)
+{
+	auto num_routes = get_unique_hosts(sess).size();
+
+	BOOST_REQUIRE_MESSAGE(num_routes == nodes_count,
+			      "Network is not convergenced: num_routes (" + std::to_string(static_cast<long long>(num_routes)) +
+			      ") != total_nodes_count (" + std::to_string(static_cast<long long>(nodes_count)) + ")");
+}
+
+bool register_tests(test_suite *suite, session &sess)
+{
+	ELLIPTICS_TEST_CASE(test_network_convergence, sess);
+
+	return true;
+}
+
+static void destroy_global_data()
+{
+	global_data.reset();
+}
+
+boost::unit_test::test_suite *register_tests(int argc, char *argv[])
+{
+	namespace bpo = boost::program_options;
+
+	bpo::variables_map vm;
+	bpo::options_description generic("Test options");
+
+	std::string path;
+
+	generic.add_options()
+			("help", "This help message")
+			("path", bpo::value(&path), "Path where to store everything")
+			;
+
+	bpo::store(bpo::parse_command_line(argc, argv, generic), vm);
+	bpo::notify(vm);
+
+	if (vm.count("help")) {
+		std::cerr << generic;
+		return NULL;
+	}
+
+	test_suite *suite = new test_suite("Local Test Suite");
+
+	configure_nodes(path);
+
+	session sess = create_session(*global_data->node, { 1 }, 0, 0);
+	// wait for network connectivity convergence between server nodes
+	sleep(nodes_count);
+
+	register_tests(suite, sess);
+
+	return suite;
+}
+
+}
+
+int main(int argc, char *argv[])
```

Comment:
We still don't know all conditions which make this bug always reproducible, but race condition catched pretty often.